File: src/velvet/main/Velvet.java

package velvet.main;

import velvet.structs.Position;
import velvet.velements.interact.UIEventHandler;

public abstract class Velvet {

    protected Mouse mouse;
    protected Keyboard keyboard;
    protected UIEventHandler uiEventHandler;
    protected FileDrop fileDrop;
    protected Position size;
    
    public Velvet(Position size){
        this.size = size;
    }
    
    public abstract void init();
    
    protected final void setMouse(Mouse mouse){ this.mouse = mouse; }
    protected final void setKeyboard(Keyboard keyboard){ this.keyboard = keyboard; }
    protected final void setUiEventHandler(UIEventHandler uiEventHandler) { this.uiEventHandler = uiEventHandler; }
    protected final void setFileDrop(FileDrop fileDrop){ this.fileDrop = fileDrop; }
    
    public abstract void update();
    public abstract void render(VGraphics g);

    public abstract void onClose();
    public final Position getSize(){ return size; }
    
    public static void start(Velvet level, String name){
        new Main(level, name);
    }
}

File: src/velvet/main/Velvet.kt
package velvet.main

import velvet.io.hardware.InputEventStream
import velvet.ui.UIEventHandler
import velvet.ui.UINode
import velvet.util.types.spatial.Size

class VelvetState(val size: Size,
                  val inputEventStream: InputEventStream,
                  val uiEventHandler: UIEventHandler,
                  val rootNode: UINode,
                  val fileDrop: FileDrop)

abstract class Velvet(private val velvetState: VelvetState){

    val size: Size get() = velvetState.size
    val inputEventStream: InputEventStream get() = velvetState.inputEventStream
    val uiEventHandler: UIEventHandler get() = velvetState.uiEventHandler
    val rootNode: UINode get() = velvetState.rootNode
    val fileDrop: FileDrop get() = velvetState.fileDrop

    abstract fun onClose()

    companion object {
        fun start(levelGenerator: (VelvetState)->Velvet, size: Size, name: String) {
            Main(levelGenerator, size, name)
        }
    }
}